### PR TITLE
Azure: Support authentication using user-assigned managed identity

### PR DESCRIPTION
**Pull request by amrmahdi**
Adds support for [user-assigned managed identity](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview#managed-identity-types).

When using the `user_assigned_id` the change does not require setting `msi_resource` it defaults to `https://<storage account name>.<endpoint>` unless the config explicitly sets the `msi_resource`.

Tested against Azure.

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

### CODE REVIEW

File: pkg/objstore/azure/helpers.go [hunk 1]
```diff
@@ -37,11 +38,27 @@ func init() {
...
+		if resource == "" {
+			resource = fmt.Sprintf("https://%s.%s", conf.StorageAccountName, conf.Endpoint)
+		}
+
+		var (
+			spt *adal.ServicePrincipalToken
+			err error
+		)
+
+		if conf.UserAssignedID != "" {
```
Comment (wiardvanrij) [inline pkg/objstore/azure/helpers.go:52]: What if there is a function like `getServicePrincipalToken` which just returns the `*adal.ServicePrincipalToken`? That removes those two vars defines and allows for nice returns.

**Review by wiardvanrij — APPROVED** [review]
Ah this is nice. Explains the different auth methods much better than I could figure out from the MS docs lol. Readme is spot on and makes a lot of sense. Just a little nitpick/suggestion on the code logic.

File: pkg/objstore/azure/helpers.go (hunk 1 above)
Comment (amrmahdi) [reply pkg/objstore/azure/helpers.go:52]: Nice suggestion

**Review by wiardvanrij — DISMISSED** [review]

**Comment by amrmahdi** [comment]
@bwplotka Hoping this can get on the 0.23 release treain :)

File: pkg/objstore/azure/helpers.go [hunk 2]
```diff
@@ -68,6 +66,23 @@ func getAzureStorageCredentials(logger log.Logger, conf Config) (blob.Credential
...
 
+func getServicePrincipalToken(logger log.Logger, conf Config) (*adal.ServicePrincipalToken, error) {
+	resource := conf.MSIResource
+	if resource == "" {
+		resource = fmt.Sprintf("https://%s.%s", conf.StorageAccountName, conf.Endpoint)
+	}
+
+	if conf.UserAssignedID != "" {
+		level.Debug(logger).Log("msg", "using user assigned identity:", conf.UserAssignedID)
+		return adal.NewServicePrincipalTokenFromMSIWithUserAssignedID("", resource, conf.UserAssignedID)
```
Comment (phillebaba) [inline pkg/objstore/azure/helpers.go:77]: Nit: This function is deprecated, instead `NewServicePrincipalTokenFromManagedIdentity` should be used.
https://github.com/Azure/go-autorest/blob/cf74520d3483ff497328aa9f0f0e2ed5e55a70ad/autorest/adal/token.go#L749-L752

Comment (amrmahdi) [reply pkg/objstore/azure/helpers.go]: Thanks for the note. Updated to not use the deprecated method. I ended up using the same constructs as the system-assigned identity existing code, and only set the client id if the user assigned identity is specified, which further simplified the code.

**Review by GiedriusS — APPROVED** [review]
:+1:

**Pull request merged**